plugins/solaris/io_disk

```diff
@@ -0,0 +1,301 @@
+. "${MUNIN_LIBDIR:-}/plugins/plugin.sh"
+is_multigraph "$@"
+
+# Shell options
+set -o nounset
+set -o pipefail
+
+# Set environment variables
+plugin_name=io
+functions='ops bytes busy queue latency'
+: "${class:=disk}"
+: "${module:=sd}"
+: "${module_regex:=/^(s?sd|vdc|zvblk|dad|md)\$/}"
+: "${title_type:=Disk Device}"
+: "${exclude:=}"
+: "${graph_width:=}"
+
+# Create map of instance name (e.g. sd0) and logical device name (e.g. c0t0d0)
+#   Example: 
+#     name_sd1=c0t0d0
+#     name_ssd2=c0tAB_1234d0  (shorten long target)
+#     ...
+declare $( paste -d= \
+    <( iostat -x | awk '{print $1}' | sed -e '1,2d' -e 's/^/name_/' ) \
+    <( iostat -xn | awk '{print $NF}' | sed -e '1,2d' -e 's/^\(c[0-9]*\)\(t.\{2\}\).*\(.\{4\}\)\(d[0-9]*\)$/\1\2_\3\4/' ) \
+    )
+
+# Functions
+
+preconfig() {
+    local func
+    func=$1
+
+    case "$func" in
+    ops)
+        conf_title='I/O Operations'
+        conf_in=reads
+        conf_out=writes
+        conf_cdef=
+        conf_gargs='--base 1000'
+        conf_vlabel='Ops per second write (-) / read (+)'
+        ;;
+    bytes)
+        conf_title='I/O Throughput'
+        conf_in=nread
+        conf_out=nwritten
+        conf_cdef=
+        conf_gargs='--base 1024'
+        conf_vlabel='Bytes per second write (-) / read (+)'
+        ;;
+    busy)
+        conf_title='Busy & Wait'
+        conf_in=rtime
+        conf_out=wtime
+        conf_cdef=',100,*'
+        conf_gargs='--base 1000 --lower-limit 0 --upper-limit 100'
+        conf_vlabel='% wait (-) / busy (+)'
+        ;;
+    queue)
+        conf_title='Queue Length'
+        conf_in=rlentime
+        conf_out=wlentime
+        conf_cdef=
+        conf_gargs='--base 1000'
+        conf_vlabel='Queue length wait (-) / actv (+)'
+        ;;
+    latency)
+        conf_title='Latency'
+        conf_in=rlentime
+        conf_out=wlentime
+        conf_cdef=
+        conf_gargs='--base 1000'
+        conf_vlabel='Seconds wsvc_t (-) / asvc_t (+)'
+        ;;
+    *)
+        echo "Unknown function: $func"
+        exit 1
+        ;;
+    esac
+}
+
+is_excluded() {
+    local dev i
+    dev=$1
+
+    for i in ${exclude}
+    do
+        if [ "$dev" = "$i" ]; then
+            return 0
+        fi
+    done
+
+    return 1
+}
+
+do_config() {
+    local func dev ref devname stat
+
+    func=$1
+    preconfig "$func"
+
+    echo "multigraph ${plugin_name}_${func}_${module}"
+
+    echo "graph_title $title_type $conf_title"
+    echo 'graph_category disk'
+    echo "graph_args $conf_gargs"
+    echo "graph_vlabel $conf_vlabel"
+    if [ -n "$graph_width" ]; then
+        echo "graph_width $graph_width"
+    fi
+
+    kstat -p -c "$class" -m "$module_regex" -s "/^${conf_in}\$/" | sed 's/:/ /g' | awk '{ print $3 }' | \
+    while read -r dev
+    do
+        if is_excluded "$dev" ; then
+            continue
+        fi
+
+        # replace instance name to logical device name if exists
+        ref="name_$dev"
+        devname=${!ref:-}
```

Comment:
Agree. I will add a comment. Indirect evaluation looks pragmatic, but it may be better than eval. 
I actually prefer hashes, but it is supported above bash 4. Solaris 11 has bash 4, but Solaris 10 has only bash 3.